File: other-programming-languages/PHP/2024/2024-005/2024-005.php
<?php

// Reading the file content
$filePath = 'input.txt';
$content = file_get_contents($filePath);

// Splitting content into rules and updates
list($rulesSection, $updatesSection) = explode("\n\n", trim($content));

$rules = array_map(function ($line) {
    return array_map('intval', explode('|', $line));
}, explode("\n", $rulesSection));

$updates = array_map(function ($line) {
    return array_map('intval', explode(',', $line));
}, explode("\n", $updatesSection));

// Helper function to check if an update follows the rules
function isUpdateOrdered($update, $rules) {
    $indexMap = array_flip($update);
    foreach ($rules as list($x, $y)) {
        if (isset($indexMap[$x], $indexMap[$y]) && $indexMap[$x] > $indexMap[$y]) {
            return false;
        }
    }
    return true;
}

// Identify correctly ordered updates and their middle page numbers
$correctUpdates = [];
$middlePages = [];

foreach ($updates as $update) {
    if (isUpdateOrdered($update, $rules)) {
        $correctUpdates[] = $update;
        $middlePages[] = $update[intval(floor(count($update) / 2))];
    }
}

// Calculate the sum of middle pages
$sumMiddlePages = array_sum($middlePages);
echo "Sum of middle pages for correctly ordered updates: $sumMiddlePages\n";

// Helper function to sort an update according to the rules
function topologicalSort($update, $rules) {
    $graph = [];
    $inDegree = [];
    $nodes = array_flip($update);

    foreach ($rules as list($x, $y)) {
        if (isset($nodes[$x], $nodes[$y])) {
            $graph[$x][] = $y;
            $inDegree[$y] = ($inDegree[$y] ?? 0) + 1;
            $inDegree[$x] = $inDegree[$x] ?? 0;
        }
    }

    $queue = [];
    foreach ($nodes as $node => $_) {
        if (($inDegree[$node] ?? 0) === 0) {
            $queue[] = $node;
        }
    }

    $sortedUpdate = [];
    while (!empty($queue)) {
        $current = array_shift($queue);
        $sortedUpdate[] = $current;

        foreach ($graph[$current] ?? [] as $neighbor) {
            $inDegree[$neighbor]--;
            if ($inDegree[$neighbor] === 0) {
                $queue[] = $neighbor;
            }
        }
    }

    return $sortedUpdate;
}

// Correct the incorrectly ordered updates and find their middle page numbers
$incorrectUpdates = [];
$incorrectMiddlePages = [];

foreach ($updates as $update) {
    if (!isUpdateOrdered($update, $rules)) {
        $correctedUpdate = topologicalSort($update, $rules);
        $incorrectUpdates[] = $correctedUpdate;
        $incorrectMiddlePages[] = $correctedUpdate[intval(floor(count($correctedUpdate) / 2))];
    }
}

// Calculate the sum of middle pages for corrected updates
$sumIncorrectMiddlePages = array_sum($incorrectMiddlePages);
echo "Sum of middle pages for corrected updates: $sumIncorrectMiddlePages\n";

?>

File: other-programming-languages/Ruby/2024/2024-005/2024-005.rb
require 'set'

# Reading the file content
file_path = 'input.txt'
content = File.read(file_path)

# Splitting content into rules and updates
rules_section, updates_section = content.strip.split("\n\n")
rules = rules_section.split("\n").map { |line| line.split("|").map(&:to_i) }
updates = updates_section.split("\n").map { |line| line.split(",").map(&:to_i) }

# Helper function to check if an update follows the rules
def update_ordered?(update, rules)
  index_map = update.each_with_index.to_h
  rules.all? do |x, y|
    !index_map.key?(x) || !index_map.key?(y) || index_map[x] <= index_map[y]
  end
end

# Identify correctly ordered updates and their middle page numbers
correct_updates = []
middle_pages = []

updates.each do |update|
  if update_ordered?(update, rules)
    correct_updates << update
    middle_pages << update[update.length / 2]
  end
end

# Calculate the sum of middle pages
sum_middle_pages = middle_pages.sum
puts "Sum of middle pages for correctly ordered updates: #{sum_middle_pages}"

# Helper function to sort an update according to the rules
def topological_sort(update, rules)
  # Build a graph and in-degree count based on the rules
  graph = Hash.new { |h, k| h[k] = [] }
  in_degree = Hash.new(0)
  nodes = update.to_set

  rules.each do |x, y|
    if nodes.include?(x) && nodes.include?(y)
      graph[x] << y
      in_degree[y] += 1
      in_degree[x] ||= 0
    end
  end

  # Perform topological sort
  queue = nodes.select { |node| in_degree[node] == 0 }
  sorted_update = []

  until queue.empty?
    current = queue.shift
    sorted_update << current

    graph[current].each do |neighbor|
      in_degree[neighbor] -= 1
      queue << neighbor if in_degree[neighbor] == 0
    end
  end

  sorted_update
end

# Correct the incorrectly ordered updates and find their middle page numbers
incorrect_updates = []
incorrect_middle_pages = []

updates.each do |update|
  unless update_ordered?(update, rules)
    corrected_update = topological_sort(update, rules)
    incorrect_updates << corrected_update
    incorrect_middle_pages << corrected_update[corrected_update.length / 2]
  end
end

# Calculate the sum of middle pages for corrected updates
sum_incorrect_middle_pages = incorrect_middle_pages.sum
puts "Sum of middle pages for corrected updates: #{sum_incorrect_middle_pages}"
